diff --git a/src/openfermion/ops/operators/symbolic_operator.py b/src/openfermion/ops/operators/symbolic_operator.py
index fbd8d6a..1c4c0c2 100644
--- a/src/openfermion/ops/operators/symbolic_operator.py
+++ b/src/openfermion/ops/operators/symbolic_operator.py
@@ -120,8 +120,6 @@ class SymbolicOperator(metaclass=abc.ABCMeta):
     __hash__ = None

     def __init__(self, term=None, coefficient=1.0):
-        if not isinstance(coefficient, COEFFICIENT_TYPES):
-            raise ValueError('Coefficient must be a numeric type. Got {}'.format(type(coefficient)))

         # Initialize the terms dictionary
         self.terms = {}
